pkgs/development/python-modules/python-doctr/default.nix

```diff
@@ -0,0 +1,173 @@
+{ lib, buildPythonPackage, fetchPypi, fetchFromGitHub, fetchurl, fetchgit, git, glibc, python3Packages }:
+
+let
+  mplcursors = buildPythonPackage rec {
+    pname = "mplcursors";
+    version = "0.3";
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "sha256-DjLBxhP4g6Q21TrWWMA3er2ErAI12UDBNeEclKG679A";
+    };
+
+    # Build-time dependencies
+    nativeBuildInputs = [ python3Packages.setuptools_scm ];
+
+    # Run-time dependencies
+    propagatedBuildInputs = [
+      python3Packages.matplotlib
+      python3Packages.pytest
+      python3Packages.weasyprint
+    ] ++ lib.optional (python3Packages.pythonOlder "3.8") python3Packages.importlib-metadata;
+
+    meta = with lib; {
+      description = "Interactive cursors for Matplotlib";
+      homepage = "https://example.com/mplcursors";
+      license = licenses.mit;
+    };
+  };
+
+  ctypesgen = buildPythonPackage rec {
+    pname = "ctypesgen";
+    version = "pypdfium2";
+
+    src = fetchFromGitHub {
+      owner = "pypdfium2-team";
+      repo = "ctypesgen";
+      rev = "pypdfium2";
+      sha256 = "sha256-klc6mouJ8w/xIgx8xmDXrui5Ebyicg++KIgr+b5ozbk=";
+    };
+
+    # Specify native build inputs
+    nativeBuildInputs = with python3Packages; [
+      setuptools
+      wheel
+      setuptools_scm
+      tomli
+    ];
+
+    buildInputs = [ glibc ];
+
+    # Custom patching steps
+    postPatch = ''
+      export SETUPTOOLS_SCM_PRETEND_VERSION=1.0.0 # fake version
+      mkdir -p dist
+    '';
+
+    # Disable checks if necessary
+    doCheck = false;
+
+    # Run-time dependencies
+    propagatedBuildInputs = [ python3Packages.wheel python3Packages.toml ];
+
+    meta = with lib; {
+      description = "Python bindings generator for C libraries";
+      homepage = "https://github.com/pypdfium2-team/ctypesgen";
+      license = licenses.mit;
+    };
+  };
+
+  pypdfium2 = buildPythonPackage rec {
+    pname = "pypdfium2";
+    version = "4.24.0";
+
+    src = fetchPypi {
+      inherit pname version;
+      sha256 = "sha256-YnBsBrxb45qnolMa+AJCBCm2xMR0mO69JSGvfpiNCEg=";
+    };
+
+    # Additional source and binary fetching
+    headers = fetchurl {
+      url = "https://pdfium.googlesource.com/pdfium/+archive/7233e99fcaeb18adbf048be2df0b1cca355abc70/public.tar.gz";
+      sha256 = "sha256-920OK/8UXrwwlf+FBrIKdTl3Q35W1li/BEpGknbtRlU=";
+    };
+
+    binaries = fetchurl {
+      url = "https://github.com/bblanchon/pdfium-binaries/releases/download/chromium%2F6124/pdfium-linux-x64.tgz";
+      sha256 = "sha256-nFIwGgpwFV31rgu6ZFZtrcAAEltBNPgoVy5hR7evbA8=";
+    };
+
+    # Patches and post-patch steps
+    patches = [ ./pypdfdfium2-get-binaries.patch ];
+
+    # Place headers and binary downloads in the expected locations
+    postPatch = ''
+      mkdir -p data/bindings/headers
+      tar -xzf ${headers} -C data/bindings/headers
+      mkdir -p data/linux_x64
+      cp ${binaries} data/linux_x64/pdfium-linux-x64.tgz
+      cp ${binaries} pdfium-linux-x64.tgz
```

Comment:
Actually an interesting approach, just filling the data cache.
If we add caches for the tarball and the refs, that might be a viable alternative to entirely external data files management - then offline callers could just hook in at any abstraction stage they like and reuse as much of our code as possible.

(I overlooked that possibility during discussion with @nh2, though shortly afterwards it came to my mind that programatically we don't strictly need the `prepared!` target and could just go through the cache instead.)

Reply:
Wait, though. Caches need version info to be correct, so it would probably add more complexity than it's worth and you might better use external data files directly.

Also, this made me realize an issue (regression) in the cache logic - existing headers can't be used without a version check, otherwise this would fail to update. I'll fix this shortly, but the fix will break the assumption made by this code.